### ISSUE TITLE
[BUG] Well-known challenge never created.

### ISSUE THREAD

**Issue by Molgaard78** [RunnerM/simply-dns-webhook#89]
**Describe the bug**
cert-manager does not create `_acme-challenge.[*]` record for the ACME challenge at `simply.com` by using simply.com implementation of `dns-01` webhook .

**Expected behavior**
_well-know record is created based on challenged received.

**Additional context**

I have a Kubernetes cluster (K3S) that I have installed by using the [quickstart guide](https://docs.k3s.io/quick-start) and then installed [cert-manager by using Helm](https://cert-manager.io/docs/installation/helm/).

I want to use your webhook, since my DNS provider is Simply.com, but unfortunately I ran into a few troubles:

First off: If I install `simply-dns-webhook` by calling:

```
helm install simply-dns-webhook simply-dns-webhook/simply-dns-webhook --version 1.5.4
```
Then I will run into a permission error when I try to issue a certificate using `ClusterIssuer`.

It appears the role that reads the secrets from the `cert-manager` namespace is stored in the **default** namespace and is therefore not allowed to read secrets from the **cert-manager** namespace.

This is easily mitigated by adding `-n cert-manager` to the `helm install ...` command. :-)

Which leads my to my real issue.

I want to use the simply.com webhook to get certificates issued by Let's Encrypt.

Therefore I have created `letsencrypt-staging.yaml` with the following content:

```lang-yaml
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: letsencrypt-staging
spec:
  acme:
    email: me@example.com
    privateKeySecretRef:
      name: letsencrypt-staging-key
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    solvers:
    - dns01:
        webhook:
          groupName: com.github.runnerm.cert-manager-simply-webhook
          solverName: simply-dns-solver
          config:
            secretName: simply-credentials
      selector:
        dnsZones:
        - 'cluster.example.com'
        - '*.cluster.example.com'
```
To generate my test certificate I have a file called `test-certificate.yaml` with the following content:

```lang-yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: test-certificate
spec:
  dnsNames:
  - test.cluster.example.com
  issuerRef:
    name: letsencrypt-staging
    kind: ClusterIssuer

  secretName: test-certificate-tls
```
I should then be able to get a certificate by first creating the ClusterIssuer with `kubectl apply -f letsencrypt-staging.yaml` and request for a certificate with `kubectl apply -f test-certificate.yaml`.

Unfortunately this does not happen because output from `kubectl get certs` reveals the following result:

```
NAME               READY   SECRET                 AGE
test-certificate   False   test-certificate-tls   52m
```

**Screenshots/Logs**

Output from `kubectl events` gives:

```
LAST SEEN   TYPE     REASON               OBJECT                                               MESSAGE
53m         Normal   Issuing              Certificate/test-certificate                         Issuing certificate as Secret does not exist
53m         Normal   Generated            Certificate/test-certificate                         Stored new private key in temporary Secret resource "test-certificate-vjnhx"
53m         Normal   Requested            Certificate/test-certificate                         Created new CertificateRequest resource "test-certificate-1"
53m         Normal   WaitingForApproval   CertificateRequest/test-certificate-1                Not signing CertificateRequest until it is Approved
53m         Normal   WaitingForApproval   CertificateRequest/test-certificate-1                Not signing CertificateRequest until it is Approved
53m         Normal   WaitingForApproval   CertificateRequest/test-certificate-1                Not signing CertificateRequest until it is Approved
53m         Normal   WaitingForApproval   CertificateRequest/test-certificate-1                Not signing CertificateRequest until it is Approved
53m         Normal   WaitingForApproval   CertificateRequest/test-certificate-1                Not signing CertificateRequest until it is Approved
53m         Normal   cert-manager.io      CertificateRequest/test-certificate-1                Certificate request has been approved by cert-manager.io
53m         Normal   OrderCreated         CertificateRequest/test-certificate-1                Created Order resource default/test-certificate-1-1965387138
53m         Normal   OrderPending         CertificateRequest/test-certificate-1                Waiting on certificate issuance from order default/test-certificate-1-1965387138: ""
53m         Normal   Created              Order/test-certificate-1-1965387138                  Created Challenge resource "test-certificate-1-1965387138-1295925723" for domain "test.cluster.example.com"
53m         Normal   Started              Challenge/test-certificate-1-1965387138-1295925723   Challenge scheduled for processing
53m         Normal   Presented            Challenge/test-certificate-1-1965387138-1295925723   Presented challenge using DNS-01 challenge mechanism
```

Output from `kubectl describe challenge test-certificate-1-1965387138-1295925723` reveals amongst others the following lines:

```
Status:
  Presented:   true
  Processing:  true
  Reason:      Waiting for DNS-01 challenge propagation: DNS record for "test.cluster.example.com" not yet propagated
  State:       pending
Events:
  Type    Reason     Age   From                     Message
  ----    ------     ----  ----                     -------
  Normal  Started    59m   cert-manager-challenges  Challenge scheduled for processing
  Normal  Presented  59m   cert-manager-challenges  Presented challenge using DNS-01 challenge mechanism
```

And when I login Simply.com website then I cannot see any records containing the word `_acme-challenge`, which is used by the ACME protocol.

I can see In your code that you handoff communication with Simply.com to `simply-com-client`, but I cannot see any debug information about whether not the client is getting called or if there is an authentication error between the client and Simply.com?

Where can I look to debug further?

**Comment by RunnerM** [comment]
Will look at it later.😉

**Comment by RunnerM** [comment]
Please take a look at the webhook's logs and maybe set it to debug mode. 😉 Also you can provide it here if you want me to take a look.

**Comment by Molgaard78** [comment]
I am quite new to kubernetes, however a little more digging by running first `kubectl get pods -n cert-manager` gave me that simply-dns-webhook is running in the pod `simply-dns-webhook-6f87bbb8db-t5njw`.

From there I delete my old test certificate then ran the `test-certificate.yaml` again, so it was easier to find in the log.

It gave me the following output from `kubectl logs simply-dns-webhook-6f87bbb8db-t5njw -n cert-manager`:

```
I0330 14:21:47.668912       1 handler.go:275] Adding GroupVersion com.github.runnerm.cert-manager-simply-webhook v1alpha1 to ResourceManager
I0330 14:21:47.677105       1 secure_serving.go:213] Serving securely on [::]:443
I0330 14:21:47.677170       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I0330 14:21:47.677178       1 shared_informer.go:311] Waiting for caches to sync for RequestHeaderAuthRequestController
I0330 14:21:47.677209       1 dynamic_serving_content.go:132] "Starting controller" name="serving-cert::/tls/tls.crt::/tls/tls.key"
I0330 14:21:47.677314       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
time="2024-03-30T14:21:47Z" level=info msg="Initializing Simply dns solver"
I0330 14:21:47.677700       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I0330 14:21:47.677711       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0330 14:21:47.677733       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I0330 14:21:47.677738       1 shared_informer.go:311] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0330 14:21:47.777289       1 shared_informer.go:318] Caches are synced for RequestHeaderAuthRequestController
I0330 14:21:47.778055       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0330 14:21:47.778060       1 shared_informer.go:318] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
time="2024-03-30T14:22:12Z" level=info msg="Presenting challenge for: _acme-challenge.test.cluster.example.com."
time="2024-03-30T14:22:13Z" level=error msg="Failed request, response code:  401"
time="2024-03-30T14:22:13Z" level=error msg="Failed request, response:  401"
time="2024-03-30T17:55:08Z" level=info msg="Cleaning up challenge for: _acme-challenge.test.cluster.example.com."
time="2024-03-30T17:55:08Z" level=error msg="Failed request, response code:  401"
time="2024-03-30T17:55:08Z" level=error msg="Failed request, response code:  401"
time="2024-03-30T17:55:08Z" level=error msg="record(0) have not been cleaned up"
time="2024-03-30T17:55:51Z" level=info msg="Presenting challenge for: _acme-challenge.test.cluster.example.com."
time="2024-03-30T17:55:51Z" level=error msg="Failed request, response code:  401"
time="2024-03-30T17:55:51Z" level=error msg="Failed request, response:  401"
```

I get that the `401` error code is an authentication error, but against what? :-)

I just checked to be sure that I stored the correct data for my Simply.com account by verifying that the account name in Simply.com's control page is exactly the same as:

```
kubectl get secrets -n cert-manager simply-credentials -ojson | jq '.data."account-name"' -r | base64 -d
```

And API key on Simply.com's control page is exactly the same as:

```
kubectl get secrets -n cert-manager simply-credentials -ojson | jq '.data."api-key"' -r | base64 -d
```

**Comment by RunnerM** [comment]
I suggest the same as in this issue from before: [here](https://github.com/RunnerM/simply-dns-webhook/issues/66) to do a test run with the config option and then try to debug your secret my guess is that your base64 conversion is wrong. I have been there and many other people did that as well, so no worries.🙂

**Comment by Molgaard78** [comment]
Strange. The way I converted was something like: `echo SECRET_KEY | base64`, which should give the correct base64 encoded secret.

But you are into something because when I encoded using https://www.base64encode.org/ I could see 1 char being replaced in `account-name` and `api-key` getting shortened by 4 chars.

I guess the devil is in the details...?

**Comment by RunnerM** [comment]
I looked up why the encoding would fail for you and the explanation is that, when you do this:  `echo SECRET_KEY | base64` you unknowingly add a new line character with the echo command at the end of your string. Which can be mitigated with the `-n` flag like so  `echo -n SECRET_KEY | base64` to produce the correct encoded string.  ;)  

Also many services packaged for K8S read configs and secrets once on startup or only deploy in some cases. With this package it is enough to restart the pod.

Hope this helps!  ;)

**Comment by Molgaard78** [comment]
It did. 👍 

You learn something new every day. I have used like two months on and off trying to learn Kubernetes and having issues like not being able to generate certificate can be a bit off putting in doing your own experments.

Particularly since many experiments require a browser and for instance the Chrome browser will at times simply not show the webpage if it is NOT protected by a certificate.

The fact that non-visible characters can affect the code you are generating is very hard to debug.

The last time I had to go debugging in the extreme was when wrote a program for a USB modem many years ago that communicated via SMS.

Text was garbled up, so I did pen and paper debugging on the encoding ... at binary level!

That is the you get when you have string consisting of 7-bit characters converted into a string of 8-bit characters and encoding used BigIndian if I remember right.
